### PR TITLE
t8n: Add tx hash calculation for tx receipt

### PR DESCRIPTION
**Pull request by rodiazet**
- Change is required by `execution-spec-tests`. The tool doesn't set `hash` field for transaction in `txs.json`. We have to calculate it.
- Add better error printing to console.

### CODE REVIEW

File: test/t8n/t8n.cpp [hunk 1]
```diff
@@ -120,7 +120,8 @@ int main(int argc, const char* argv[])
 
                         txs_logs.insert(txs_logs.end(), tx_logs.begin(), tx_logs.end());
                         auto& j_receipt = j_result["receipts"][j_result["receipts"].size()];
-                        j_receipt["transactionHash"] = j_txs[i]["hash"];
```
[the remark this answers is not in the document]
Comment (axic) [reply test/t8n/t8n.cpp:123]: `j_rejected_tx` above also uses this field, probably need to change that too?

File: test/t8n/t8n.cpp [hunk 2]
```diff
@@ -165,9 +166,9 @@ int main(int argc, const char* argv[])
 
         std::ofstream{output_dir / output_alloc_file} << j_alloc;
     }
-    catch (...)
+    catch (const std::exception& e)
     {
-        std::cerr << "Unhandled exception" << std::endl;
+        std::cerr << e.what() << std::endl;
```
Comment (chfast) [inline test/t8n/t8n.cpp:171]: I don't think we should exit with `0` in this case.

File: test/t8n/t8n.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (chfast) [reply test/t8n/t8n.cpp:123]: The `"hash"` was added by our request. But it is fine to compute it now.
If `"hash"` is available we should compare it with the computed hash and throw an exception if mismatch.

**Review by chfast — CHANGES_REQUESTED** [review]

File: test/t8n/t8n.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (rodiazet) [reply test/t8n/t8n.cpp:123]: done

File: test/t8n/t8n.cpp [hunk 3]
```diff
@@ -103,11 +103,16 @@ int main(int argc, const char* argv[])
                     tx.chain_id = chain_id;
 
                     auto res = state::transition(state, block, tx, rev, vm);
+
+                    const auto tx_hash = hex0x(keccak256(rlp::encode(tx)));
+                    if (j_txs[i].contains("hash") && j_txs[i]["hash"] != tx_hash)
```
Comment (chfast) [inline test/t8n/t8n.cpp:108]: Comparing string is risky. They may differ by "0x" prefix or hex digit case. Keep them as `hash256` type and only convert to string for printing.

File: test/t8n/t8n.cpp [hunk 4]
```diff
@@ -103,11 +103,16 @@ int main(int argc, const char* argv[])
...
+                        throw std::invalid_argument("hash mismatch");
```
Comment (chfast) [inline test/t8n/t8n.cpp:109]: Provide more information: computed hash, loaded hash, "transaction hash".
And maybe use some other `logic_error` exception. `std::domain_error`?

**Review by axic — APPROVED** [review]
Looks good to me otherwise.

**Review by chfast — CHANGES_REQUESTED** [review]

File: test/t8n/t8n.cpp [hunk 5]
```diff
@@ -103,11 +103,16 @@ int main(int argc, const char* argv[])
...
+
                     if (holds_alternative<std::error_code>(res))
                     {
                         const auto ec = std::get<std::error_code>(res);
                         json::json j_rejected_tx;
-                        j_rejected_tx["hash"] = j_txs[i]["hash"];
+                        j_rejected_tx["hash"] = tx_hash;
```
[the remark this answers is not in the document]
Comment (rodiazet) [reply test/t8n/t8n.cpp:115]: `hash` https://github.com/ethereum/retesteth/blob/master/retesteth/testStructures/types/RPC/SubElements/ToolResponseRejected.cpp#L13

File: test/t8n/t8n.cpp (hunk 3 above)
Comment (rodiazet) [reply test/t8n/t8n.cpp:108]: Done

File: test/t8n/t8n.cpp (hunk 4 above)
Comment (rodiazet) [reply test/t8n/t8n.cpp:109]: Done

File: test/t8n/t8n.cpp [hunk 6]
```diff
@@ -103,11 +103,24 @@ int main(int argc, const char* argv[])
                     tx.chain_id = chain_id;
 
                     auto res = state::transition(state, block, tx, rev, vm);
+
+                    const auto tx_computed_hash = keccak256(rlp::encode(tx));
```
Comment (chfast) [inline test/t8n/t8n.cpp:107]:
```suggestion
                    const auto computed_tx_hash = keccak256(rlp::encode(tx));
```
Keep "tx hash" words together.

File: test/t8n/t8n.cpp [hunk 7]
```diff
@@ -103,11 +103,24 @@ int main(int argc, const char* argv[])
...
+
+                    if (j_txs[i].contains("hash") && j_txs[i]["hash"].is_string())
```
Comment (chfast) [inline test/t8n/t8n.cpp:109]:
```suggestion
                    if (j_txs[i].contains("hash"))
```
Too defensive.

**Review by chfast — CHANGES_REQUESTED** [review]

**Review by chfast — APPROVED** [review]

File: test/t8n/t8n.cpp [hunk 8]
```diff
@@ -103,11 +103,25 @@ int main(int argc, const char* argv[])
...
+
+                    if (j_txs[i].contains("hash"))
+                    {
+                        const auto loaded_tx_hash_opt =
+                            evmc::from_hex<bytes32>(j_txs[i]["hash"].get<std::string>());
+
+                        if (loaded_tx_hash_opt != computed_tx_hash)
+                            throw std::logic_error("transaction hash mismatched: computed " +
+                                                   hex0x(computed_tx_hash) + ", expected " +
+                                                   hex0x(loaded_tx_hash_opt.value()));
```
[the remark this answers is not in the document]
Comment (rodiazet) [reply test/t8n/t8n.cpp:117]: Merged. I will add new PR.

**Pull request merged**